app/services/subscriptions/terminate_service.rb

```diff
@@ -111,17 +111,31 @@ def bill_subscription
           subscription.terminated_at,
           invoicing_reason: :subscription_terminating
         )
-        BillNonInvoiceableFeesJob.set(wait: 2.seconds).perform_later([subscription], subscription.terminated_at)
+        BillNonInvoiceableFeesJob.set(wait: 2.seconds).perform_later(
+          invoiceable_subscriptions.to_a,
+          subscription.terminated_at
+        )
       else
         BillSubscriptionJob.perform_now(
           [subscription],
           subscription.terminated_at,
           invoicing_reason: :subscription_terminating
         )
-        BillNonInvoiceableFeesJob.perform_now([subscription], subscription.terminated_at)
+        BillNonInvoiceableFeesJob.perform_now(
+          invoiceable_subscriptions.to_a,
+          subscription.terminated_at
+        )
       end
     end
 
+    def invoiceable_subscriptions
+      # Expect all subscriptions belogns to the same organization
+      @invoiceable_subscriptions ||= subscription.organization.subscriptions.where(
```

Comment:
This might return multiple terminated subscriptions for this external_id, is that expected?

Reply:
yes, this is the fix...

for non invoiceable fees we need to wait until the is payment succeeded to generate the invoice. This could happen at whatever time after the fee was created (for example after a few months or even years). The invoice will be generated once the fee is payment succeeded and the subscription is billed because of termination, deletion or end of billing period.

The scenario we are fixing here is....

1. Subscription A is created.
2. A fee is generated.
3. Subscription B upgrades subscription A
4. Fee is marked as payment succeeded
5. Subscription B is terminated or end of billing period is reached

At this point the fee should be invoiced however the fee belongs to the subscription A, not the subscription being terminated at this time that is subscription  B.

This finds all related subscriptions to the subscription being terminated and generates invoices for all non invoiceable fees....

Also, writing this comment I realised the scenario of the end of billing period should be checked  as this is only about termination.

Reply:
yes, but you can also have Subscription C, D and E that were already terminated. this query returns all of those, no?